File: solutions/binary-prefix-divisible-by-5/binary-prefix-divisible-by-5.cpp
// @Title: Binary Prefix Divisible By 5
// @Author: realzhangm
// @Date: 2021-10-29T17:02:33+08:00
// @URL: https://leetcode-cn.com/problems/binary-prefix-divisible-by-5


class Solution {
public:
    vector<bool> prefixesDivBy5(vector<int>& nums) {
        int size = nums.size();
        vector<bool> res(size, false);
        int num = 0;
        for (int i = 0; i < size; i++) {
            num = (num << 1) + nums[i];
            num = num % 5;
            res[i] = (num == 0);
        }
        return res;
    }
};

File: solutions/binary-prefix-divisible-by-5/binary-prefix-divisible-by-5.java
// @Title: Binary Prefix Divisible By 5
// @Author: colinjxc
// @Date: 2021-10-27T23:42:39+08:00
// @URL: https://leetcode-cn.com/problems/binary-prefix-divisible-by-5


class Solution {
    // 有限状态机 DFA ？
    public List<Boolean> prefixesDivBy5(int[] nums) {
        List<Boolean> res = new ArrayList<>();
        if (nums.length < 1) {
            return res;
        }
        int StateSet[][] = new int[][]{
                {0, 1},
                {2, 3},
                {4, 0},
                {1, 2},
                {3, 4}
        };
        int state = 0;
        for (int i = 0; i < nums.length; i++) {
            state = StateSet[state][nums[i]];  //转换后的状态
            if (state == 0) {
                res.add(Boolean.TRUE);
            } else {
                res.add(Boolean.FALSE);
            }
        }
        return res;
    }
}
